services/QuillLMS/lib/tasks/diagnostic_question_optimal_concepts.rake

```diff
@@ -0,0 +1,156 @@
+    "Intermediate Post" => 1669,
+    "Advanced Pre" => 1678,
+    "Advanced Post" => 1680,
+    "ELL Starter Pre" => 1161,
+    "ELL Starter Post" => 1774,
+    "ELL Intermediate Pre" => 1568,
+    "ELL Intermediate Post" => 1814,
+    "ELL Advanced Pre" => 1590,
+    "ELL Advanced Post" => 1818,
+    "Springboard" => 1432,
+    "PreAP 1" => 1229,
+    "PreAP 2" => 1230,
+    "AP" => 992
+  }
+
+  desc 'Populate diagnostic_question_optimal_concepts from CSV extracted from Curriculum Notion database: https://www.notion.so/quill/f937b1cc8d2d4ed8943a36ce35dca177?v=7dd55ee1f04e4b5daf79645aae960aa1'
+  task populate_from_csv: :environment do
+    pipe_data = $stdin.read unless $stdin.tty?
+
+    unless pipe_data
+      puts 'No data detected on STDIN.  You must pass data to the task for it to run.  Example:'
+      puts '  rake diagnostic_question_optimal_concepts:populate_from_csv < path/to/local/file.csv'
+      puts ''
+      puts 'If you are piping data into Heroku, you need to include the --no-tty flag:'
+      puts '  heroku run rake diagnostic_question_optimal_concepts:populate_from_csv -a empirical-grammar --no-tty < path/to/local/file.csv'
+      exit 1
+    end
+
+    diagnostics = DIAGNOSTIC_ID_LOOKUP.transform_values {|id| Activity.find(id)}
+
+    total_count = 0
+    duplicate_count = 0
+    missing_concept_count = 0
+    CSV.parse(pipe_data, headers: true).each do |row|
+      activity = diagnostics[row[DIAGNOSTIC_INDEX]]
+      activity_questions = activity.data['questions'].filter {|d| d['questionType'] != 'titleCards' }
+      question_number = row[QUESTION_NUMBER_INDEX].to_i
+      question_uid = activity_questions[question_number - 1]['key']
+      question = Question.find_by(uid: question_uid)
+      concepts_row = row[OPTIMAL_CONCEPTS_INDEX]
+
+      full_concept_names = concepts_row.split(DELIMITER).reject{|concept| concept == ""}.uniq
+      concepts = full_concept_names.map do |concept_name|
+        level0, level1, level2 = concept_name.split("|").map(&:strip).reverse
+        concepts = Concept.left_outer_joins(:parent)
+          .joins("LEFT OUTER JOIN concepts AS grandparent ON parent.parent_id = grandparent.id")
+          .where(visible: true)
+          .where.not(parent: {parent_id: nil})
+        concepts = concepts.where(parent: {name: level1}) if level1
+        concepts = concepts.where(grandparent: {name: level2}) if level2
+        concept = concepts.where(name: level0)
+        if concept.length > 1
+          puts "Matched too many Concepts: #{concept_name}"
+        elsif concept.length == 1
+          # A number of questions are used in multiple diagnostics, but should have the same Optimal Concepts for each question.  Since each question has a row in the CSV for every diagnostic it's in, we expect to have about 50 duplicates which we can skip over
+          begin
+            DiagnosticQuestionOptimalConcept.create!(concept_id: concept.first.id, question_uid: question.uid)
```

Comment:
If you use `create_or_update!` here, this script will be more reusable in the future - if Curriculum sends us an edited CSV, we can just run the script again and the updates will persist.

Reply:
I'm not sure that `create_or_update!` is actually different from `create!`.  Because the index is on the combo-key of `concept_id` and `question_uid`, if there's a record with both of those values in a later update, it's just the same value.  There's nothing to update, right?  Either the record with that pair exists or it doesn't.  Maybe I'm missing something, though?

Reply:
You're right, given that the only data persisted is the data within the index keys.